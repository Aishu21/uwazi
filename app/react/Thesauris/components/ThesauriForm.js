import React, {Component, PropTypes} from 'react';
import {bindActionCreators} from 'redux';
import {reduxForm} from 'redux-form';
import {Link} from 'react-router';
import 'app/Thesauris/scss/thesauris.scss';

import {saveThesauri, resetThesauri} from 'app/Thesauris/actions/thesauriActions';

export class ThesauriForm extends Component {

  componentWillUnmount() {
    this.props.resetThesauri();
  }

  render() {
    const {fields, handleSubmit, submitFailed} = this.props;

    return (
      <div className="row thesauri">
        <main className="col-sm-12">
          <div className="well thesauri">
            <form onSubmit={handleSubmit(this.props.saveThesauri)} >
              <div className="thesauri-buttons">
                <Link to="/metadata" className="btn btn-default"><i className="fa fa-arrow-left"></i> Back</Link>&nbsp;
                <button className="btn btn-success save-template">
                  <i className="fa fa-save"/> Save Thesauri
                </button>
              </div>
              <div className={'form-group thesauri-name' + (submitFailed && fields.name.invalid ? ' has-error' : '')}>
                <label htmlFor="thesauriName" className="control-label">Thesauri name</label>
                <input id="thesauriName" className="form-control" type="text" {...fields.name}/>
              </div>
              <div>Values</div>
              <div className="thesauri-values">
              {fields.values.map((value, index) => {
                return <div key={index} className="form-group">
                        <input className="form-control" type="text" {...value.label} />
                        <a className="btn btn-danger" onClick={() => fields.values.removeField(index)}>Delete</a>
                      </div>;
              })}
              </div>
            </form>
            <button onClick={() => fields.values.addField()} className="btn btn-success"><i className="fa fa-plus"></i>Add value</button>
          </div>
        </main>
      </div>
    );
  }
}

ThesauriForm.propTypes = {
  fields: PropTypes.object.isRequired,
  saveThesauri: PropTypes.func,
  resetThesauri: PropTypes.func,
  values: PropTypes.object,
  handleSubmit: PropTypes.func,
  submitFailed: PropTypes.bool
};

const validate = (values) => {
  let errors = {};

  if (!values.name) {
    errors.name = 'Required';
  }

  return errors;
};

export function mapStateToProps(state) {
  let fields = ['name', 'values[].label', 'values[].id', '_id', '_rev'];
  return {
    fields: fields,
    initialValues: state.thesauri.toJS(),
    validate,
    onSubmit: saveThesauri
  };
}

function bindActions(dispatch) {
  return bindActionCreators({saveThesauri, resetThesauri}, dispatch);
}

let form = reduxForm({form: 'thesauri'}, mapStateToProps, bindActions)(ThesauriForm);

export default form;
