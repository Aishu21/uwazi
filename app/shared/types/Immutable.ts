/** @format */

export type IImmutable<T> = T extends string
  ? string
  : T extends number
  ? number
  : T extends boolean
  ? boolean
  : T extends null
  ? null
  : T extends undefined
  ? undefined
  : T extends Array<infer Elem>
  ? {
      toJS(): Elem[];
      get(i: number): IImmutable<Elem>;
    }
  : {
      toJS(): T;
      get<Field extends keyof T>(_field: Field): IImmutable<T[Field]>;
    };
