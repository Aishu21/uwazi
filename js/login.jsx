var React = require('react');
var Router = require('react-router');

module.exports = React.createClass({

  render: function () {
    return (
      <form>
        <label>username</label>
        <input type="text"/>
      </form>
    );
  }

});
