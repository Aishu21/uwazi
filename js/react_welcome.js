/** @jsx React.DOM */

var React = require('react/addons');

/* create factory with griddle component */

var ReactApp = React.createClass({

      render: function () {

        return (
          <span>Hello World!</span>
        )
      }

  });

/* Module.exports instead of normal dom mounting */
module.exports = ReactApp;
/* Normal mounting happens inside of /main.js and is bundled with browserify */
