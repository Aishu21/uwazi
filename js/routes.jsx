var React = require("react");

var ReactRouter = require('react-router');
var Router = ReactRouter.Router;
var Route = ReactRouter.Route;

// Components
var layout = require("./layout.jsx");
var login = require("./login.jsx");

var routes = (
  <Router>
    <Route path="/" component={layout}>
      <Route path="/login" component={login}/>
    </Route>
  </Router>
);

module.exports = routes;
