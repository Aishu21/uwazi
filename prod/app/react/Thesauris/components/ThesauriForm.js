"use strict";Object.defineProperty(exports, "__esModule", { value: true });exports.mapStateToProps = mapStateToProps;exports.default = exports.ThesauriForm = void 0;var _reactReduxForm = require("react-redux-form");
var _redux = require("redux");
var _reactRedux = require("react-redux");
var _propTypes = _interopRequireDefault(require("prop-types"));
var _react = _interopRequireWildcard(require("react"));

var _Layout = require("../../Layout");
var _DragAndDrop = require("../../Layout/DragAndDrop");
var _UI = require("../../UI");
var _validator = require("../../Metadata/helpers/validator");
var _thesauriActions = require("../actions/thesauriActions");
var _FormGroup = _interopRequireDefault(require("../../DocumentForm/components/FormGroup"));
var _ShowIf = _interopRequireDefault(require("../../App/ShowIf"));

var _ThesauriFormItem = _interopRequireDefault(require("./ThesauriFormItem"));function _interopRequireWildcard(obj) {if (obj && obj.__esModule) {return obj;} else {var newObj = {};if (obj != null) {for (var key in obj) {if (Object.prototype.hasOwnProperty.call(obj, key)) {var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {};if (desc.get || desc.set) {Object.defineProperty(newObj, key, desc);} else {newObj[key] = obj[key];}}}}newObj.default = obj;return newObj;}}function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}function _extends() {_extends = Object.assign || function (target) {for (var i = 1; i < arguments.length; i++) {var source = arguments[i];for (var key in source) {if (Object.prototype.hasOwnProperty.call(source, key)) {target[key] = source[key];}}}return target;};return _extends.apply(this, arguments);}var REACT_ELEMENT_TYPE;function _jsx(type, props, key, children) {if (!REACT_ELEMENT_TYPE) {REACT_ELEMENT_TYPE = typeof Symbol === "function" && Symbol["for"] && Symbol["for"]("react.element") || 0xeac7;}var defaultProps = type && type.defaultProps;var childrenLength = arguments.length - 3;if (!props && childrenLength !== 0) {props = { children: void 0 };}if (props && defaultProps) {for (var propName in defaultProps) {if (props[propName] === void 0) {props[propName] = defaultProps[propName];}}} else if (!props) {props = defaultProps || {};}if (childrenLength === 1) {props.children = children;} else if (childrenLength > 1) {var childArray = new Array(childrenLength);for (var i = 0; i < childrenLength; i++) {childArray[i] = arguments[i + 3];}props.children = childArray;}return { $$typeof: REACT_ELEMENT_TYPE, type: type, key: key === undefined ? null : '' + key, ref: null, props: props, _owner: null };}

function sanitizeThesauri(thesaurus) {
  const sanitizedThesauri = Object.assign({}, thesaurus);
  sanitizedThesauri.values = sanitizedThesauri.values.
  filter(value => value.label).
  filter(value => !value.values || value.values.length).
  map(value => {
    const _value = Object.assign({}, value);
    if (_value.values) {
      _value.values = _value.values.filter(_v => _v.label);
    }
    return _value;
  });
  return sanitizedThesauri;
}

class ThesauriForm extends _react.Component {
  static validation(thesauris, id) {
    return {
      name: {
        duplicated: val => !thesauris.find(thesauri => thesauri.type !== 'template' &&
        thesauri._id !== id &&
        thesauri.name.trim().toLowerCase() === val.trim().toLowerCase()),
        required: _validator.notEmpty } };


  }

  constructor(props) {
    super(props);
    this.save = this.save.bind(this);
    this.renderItem = this.renderItem.bind(this);
    this.onChange = this.onChange.bind(this);
    this.onImportClicked = this.onImportClicked.bind(this);
    this.import = this.import.bind(this);
    this.fileInputRef = _react.default.createRef();
    this.fileFormRef = _react.default.createRef();
  }

  componentWillMount() {
    this.firstLoad = true;
  }


  componentWillReceiveProps(props) {
    props.thesauri.values.forEach((value, index) => {
      if (value.values && (!value.values.length || value.values[value.values.length - 1].label !== '')) {
        props.addValue(index);
      }
    });

    if (!props.thesauri.values.length || props.thesauri.values[props.thesauri.values.length - 1].label !== '') {
      props.addValue();
    }
  }

  componentDidUpdate(previousProps) {
    if (this.firstLoad) {
      this.firstLoad = false;
      return;
    }
    const { values } = this.props.thesauri;
    const previousValues = previousProps.thesauri.values;
    const addedValue = values.length > previousProps.thesauri.values.length;
    const lasValueIsGroup = values.length && values[values.length - 1].values;
    const previousLasValueWasGroup = previousValues.length && previousValues[previousValues.length - 1].values;
    if (lasValueIsGroup && (!previousLasValueWasGroup || addedValue)) {
      this.groups[this.groups.length - 1].focus();
    }
  }

  componentWillUnmount() {
    this.props.resetForm('thesauri.data');
    this.props.setInitial('thesauri.data');
  }

  onChange(values, groupIndex) {
    this.props.updateValues(values, groupIndex);
  }

  onImportClicked() {
    this.fileInputRef.current.click();
  }

  import() {
    const file = this.fileInputRef.current.files[0];
    this.fileFormRef.current.reset();
    const thes = sanitizeThesauri(this.props.thesauri);
    if (file) {
      this.props.importThesauri(thes, file);
    }
  }

  save(thesauri) {
    const sanitizedThesauri = sanitizeThesauri(thesauri);
    this.props.saveThesauri(sanitizedThesauri);
  }

  renderItem(value, index) {
    return (
      _react.default.createElement(_ThesauriFormItem.default, {
        ref: f => this.groups.push(f),
        value: value,
        index: index,
        removeValue: this.props.removeValue,
        onChange: this.onChange }));


  }

  render() {
    const isNew = this.props.new;
    const id = this.props.thesauri._id;
    const { values } = this.props.thesauri;
    if (!isNew && !id) {
      return false;
    }
    this.groups = [];
    return (
      _jsx("div", { className: "thesauri" }, void 0,
      _jsx(_reactReduxForm.Form, {
        model: "thesauri.data",
        onSubmit: this.save,
        validators: ThesauriForm.validation(this.props.thesauris.toJS(), this.props.thesauri._id) }, void 0,

      _jsx("div", { className: "panel panel-default thesauri" }, void 0,
      _jsx("div", { className: "panel-heading" }, void 0,
      _react.default.createElement(_FormGroup.default, _extends({}, this.props.state.name, { submitFailed: this.props.state.submitFailed }),
      _jsx(_reactReduxForm.Field, { model: ".name" }, void 0,
      _jsx("input", { id: "thesauriName", className: "form-control", type: "text", placeholder: "Thesauri name" }),
      _jsx(_ShowIf.default, { if: this.props.state.$form.touched && this.props.state.name && this.props.state.name.errors.duplicated }, void 0,
      _jsx("div", { className: "validation-error" }, void 0,
      _jsx(_UI.Icon, { icon: "exclamation-triangle", size: "xs" }), " Duplicated name"))))),





      _jsx("div", { className: "thesauri-values" }, void 0,
      _jsx("div", { className: "" }, void 0,
      _jsx("b", {}, void 0, "Items:")),

      _jsx(_DragAndDrop.DragAndDropContainer, {
        onChange: this.onChange,
        renderItem: this.renderItem,
        items: values,
        iconHandle: true })),


      _jsx("div", { className: "settings-footer" }, void 0,
      _jsx(_Layout.BackButton, { to: "/settings/dictionaries" }),
      _jsx("a", { className: "btn btn-primary", onClick: this.props.addGroup }, void 0,
      _jsx(_UI.Icon, { icon: "plus" }),
      _jsx("span", { className: "btn-label" }, void 0, "Add group")),

      _jsx("a", { className: "btn btn-primary", onClick: this.props.sortValues }, void 0,
      _jsx(_UI.Icon, { icon: "sort-alpha-down" }),
      _jsx("span", { className: "btn-label" }, void 0, "Sort")),

      _jsx("button", { type: "button", className: "btn btn-primary import-template", onClick: this.onImportClicked }, void 0,
      _jsx(_UI.Icon, { icon: "upload" }),
      _jsx("span", { className: "btn-label" }, void 0, "Import")),

      _jsx("button", { type: "submit", className: "btn btn-success save-template" }, void 0,
      _jsx(_UI.Icon, { icon: "save" }),
      _jsx("span", { className: "btn-label" }, void 0, "Save"))))),




      _react.default.createElement("form", { ref: this.fileFormRef, style: { display: 'none' } },
      _react.default.createElement("input", {
        ref: this.fileInputRef,
        type: "file",
        accept: "text/csv",
        style: { display: 'none' },
        onChange: this.import }))));




  }}exports.ThesauriForm = ThesauriForm;


ThesauriForm.defaultProps = {
  new: false };


















function mapStateToProps(state) {
  return {
    thesauri: state.thesauri.data,
    thesauris: state.thesauris,
    state: state.thesauri.formState };

}

function bindActions(dispatch) {
  return (0, _redux.bindActionCreators)({
    saveThesauri: _thesauriActions.saveThesauri,
    importThesauri: _thesauriActions.importThesauri,
    addValue: _thesauriActions.addValue,
    addGroup: _thesauriActions.addGroup,
    sortValues: _thesauriActions.sortValues,
    removeValue: _thesauriActions.removeValue,
    updateValues: _thesauriActions.updateValues,
    resetForm: _reactReduxForm.actions.reset,
    setInitial: _reactReduxForm.actions.setInitial,
    validate: _reactReduxForm.actions.validate },
  dispatch);
}var _default =

(0, _reactRedux.connect)(mapStateToProps, bindActions, null, { withRef: true })(ThesauriForm);exports.default = _default;