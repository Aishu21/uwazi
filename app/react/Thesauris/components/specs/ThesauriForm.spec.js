import React from 'react';
import {shallow} from 'enzyme';
import Immutable from 'immutable';

import {mapStateToProps, ThesauriForm} from '~/Thesauris/components/ThesauriForm.js';

describe('ThesauriForm', () => {
  let props;
  let component;
  beforeEach(() => {
    props = {
      fields: {name: {}, values: []},
      resetThesauri: jasmine.createSpy('resetThesauri')
    };

    component = shallow(<ThesauriForm {...props}/>);
  });

  describe('when unmount', () => {
    it('shoould call resetThesauri', () => {
      component.unmount();
      expect(props.resetThesauri).toHaveBeenCalled();
    });
  });

  describe('mapStateToProps', () => {
    let state;
    beforeEach(() => {
      state = {
        thesauri: Immutable.fromJS({name: 'thesauri name', values: []})
      };
    });

    it('should map the thesauri to initialValues', () => {
      expect(mapStateToProps(state).initialValues).toEqual({name: 'thesauri name', values: []});
    });

    it('should map the fields', () => {
      expect(mapStateToProps(state).fields).toEqual(['name', 'values[].label', 'values[].id', '_id', '_rev']);
    });
  });
});
