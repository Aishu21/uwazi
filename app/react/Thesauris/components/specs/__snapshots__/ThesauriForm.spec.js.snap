// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThesauriForm render should render groups and values 1`] = `
<div
  className="thesauri"
>
  <Connect(Form)
    model="thesauri.data"
    onSubmit={[Function]}
    validators={
      Object {
        "name": Object {
          "duplicated": [Function],
          "required": [Function],
        },
      }
    }
  >
    <div
      className="panel panel-default thesauri"
    >
      <div
        className="panel-heading"
      >
        <FormGroup>
          <Modeled(Connect(Field))
            model=".name"
          >
            <input
              className="form-control"
              id="thesauriName"
              placeholder="Thesauri name"
              type="text"
            />
            <ShowIf>
              <div
                className="validation-error"
              >
                <Component
                  icon="exclamation-triangle"
                  size="xs"
                />
                 Duplicated name
              </div>
            </ShowIf>
          </Modeled(Connect(Field))>
        </FormGroup>
      </div>
      <ul
        className="thesauri-values list-group"
      >
        <li
          className="list-group-item"
        >
          <b>
            Items:
          </b>
          <button
            alt="move"
            className="rounded-icon-small"
            onClick={[Function]}
            tabIndex={500}
            type="button"
          >
            <Connect(DirectionAwareIcon)
              icon="arrow-left"
              size="xs"
            />
          </button>
        </li>
        <ThesauriFormField
          index={0}
          key="0"
          moveToGroup={[Function]}
          moving={false}
          removeValue={[Function]}
          toggleToMove={[Function]}
          value={
            Object {
              "label": "Heroes",
              "values": Array [
                Object {
                  "label": "Batman",
                },
                Object {
                  "label": "Robin",
                },
              ],
            }
          }
        />
        <ThesauriFormField
          index={1}
          key="1"
          moveToGroup={[Function]}
          moving={false}
          removeValue={[Function]}
          toggleToMove={[Function]}
          value={
            Object {
              "label": "Villains",
              "values": Array [
                Object {
                  "label": "Joker",
                },
                Object {
                  "label": "Penguin",
                },
              ],
            }
          }
        />
      </ul>
      <div
        className="settings-footer"
      >
        <Connect(I18NLink)
          className="btn btn-default"
          to="/settings/dictionaries"
        >
          <Connect(DirectionAwareIcon)
            icon="arrow-left"
          />
          <span
            className="btn-label"
          >
            Back
          </span>
        </Connect(I18NLink)>
        <a
          className="btn btn-primary"
          onClick={[Function]}
        >
          <Component
            icon="plus"
          />
          <span
            className="btn-label"
          >
            Add group
          </span>
        </a>
        <a
          className="btn btn-primary"
          onClick={[Function]}
        >
          <Component
            icon="sort-alpha-down"
          />
          <span
            className="btn-label"
          >
            Sort
          </span>
        </a>
        <button
          className="btn btn-success save-template"
        >
          <Component
            icon="save"
          />
          <span
            className="btn-label"
          >
            Save
          </span>
        </button>
      </div>
    </div>
  </Connect(Form)>
</div>
`;
